src/persist/src/storage.rs

```diff
@@ -141,14 +141,41 @@ pub enum Atomicity {
     AllowNonAtomic,
 }
 
-/// An abstraction over a `bytes key`->`bytes value` store.
+/// An abstraction over read-only access to a `bytes key`->`bytes value` store.
+#[async_trait]
+pub trait BlobRead: Send + 'static {
+    /// Returns a reference to the value corresponding to the key.
+    async fn get(&self, key: &str) -> Result<Option<Vec<u8>>, Error>;
+
+    /// List all of the keys in the map.
+    async fn list_keys(&self) -> Result<Vec<String>, Error>;
+}
+
+/// An abstraction over read-write access to a `bytes key`->`bytes value` store.
 ///
 /// - Invariant: Implementations are responsible for ensuring that they are
 ///   exclusive writers to this location.
 #[async_trait]
-pub trait Blob: Send + 'static {
-    /// Returns a reference to the value corresponding to the key.
-    async fn get(&self, key: &str) -> Result<Option<Vec<u8>>, Error>;
+pub trait Blob: BlobRead + Sized {
+    /// The configuration necessary to open this type of storage.
+    type Config;
+    /// The corresponding [BlobRead] implementation.
+    type Read: BlobRead;
+
+    /// Opens the given location for exclusive read-write access.
+    ///
+    /// Implementations are responsible for storing the given LockInfo in such a
+    /// way that no other calls to open_exclusive succeed before this one has
+    /// been closed. However, it must be possible to continue to open this
+    /// location for reads via [Blob::open_read].
+    fn open_exclusive(config: Self::Config, lock_info: LockInfo) -> Result<Self, Error>;
+
+    /// Opens the given location for non-exclusive read-only access.
+    ///
+    /// Implementations are responsible for ensuring that this works regardless
+    /// of whether anyone has opened the same location is opened for exclusive
+    /// read-write access.
+    fn open_read(config: Self::Config) -> Result<Self::Read, Error>;
```

Comment:
sounds like we're not quite on the same page on all this. i've added a couple quick responses below, but we should probably sync up over a call

> but it feels like we specifically don't want the read-only parts of the code to import the general Blob trait

fwiw, I personally consider this a non-goal. it's not like we'll be able to save any build deps or anything

> Overall I think this function makes more sense on the BlobRead trait

this is doable. we could move `open_read` (and thus necessarily also `Config`) to BlobRead and then have Blob use `<Self as BlobRead>::Config` as the input type for `open_exclusive`. it very much feels right to me that the two methods take the same config and open_exclusive simply also takes in a LockInfo

Reply:
ack. let me read through your other PR first - I think that might help me get on the same page.

Reply:
> this is doable. we could move open_read (and thus necessarily also Config) to BlobRead and then have Blob use <Self as BlobRead>::Config as the input type for open_exclusive.

hmm maybe not. I just tried this and, since `trait Blob: BlobRead` it's unclear what open_read should return for the BlobRead impl we need to be able to impl Blob for S3Blob. it would naturally be `Self` but we don't want to return a concrete `S3Blob` from the open_read method of `impl BlobRead for S3Blob`. I guess we could introduce a third trait, something like `trait BlobRead { fn get, fn list_keys }`, `trait BlobReadConstruct: BlobRead { type Config, fn open_read }`, and `trait Blob: BlobRead { type Read: BlobReadConstruct, fn open_exclusive, fn set, ... }`. seems pretty clunky though, and I'm not yet convinced it's more clear that what this PR currently does. i'll keep mulling on it...

Reply:
discussed this offline and decided to leave things as-is. part of the confusion was an `fn as_read(&self) -> &Self::Read` that ruchir and I discussed while I was writing this PR, but was cut before I sent it out for review